pkg/plugins/shell/main_test.go

```diff
@@ -0,0 +1,53 @@
+package shell
+
+import "fmt"
+
+////////////// Test Utilities method aimed at mocking calls to other objects
+
+// mocking commandExecutor.ExecuteCommand
+type mockCommandExecutor struct {
+	gotCommand command
+	result     commandResult
+	err        error
+}
+
+func (mce *mockCommandExecutor) ExecuteCommand(cmd command) (commandResult, error) {
+	mce.gotCommand = cmd
+	if mce.gotCommand.Cmd == "" {
+		return commandResult{}, fmt.Errorf(ErrEmptyCommand)
+	}
+	return mce.result, mce.err
+}
+
+// mocking SCM object (no introspection: only get values)
+type mockScm struct {
+	workingDir string
+}
+
+func (m *mockScm) Add(files []string) error {
+	return nil
+}
+func (m *mockScm) Clone() (string, error) {
+	return "", nil
+}
+func (m *mockScm) Checkout() error {
+	return nil
+}
+func (m *mockScm) GetDirectory() (directory string) {
+	return m.workingDir
+}
+func (m *mockScm) Init(source string, pipelineID string) error {
+	return nil
+}
+func (m *mockScm) Push() error {
+	return nil
+}
+func (m *mockScm) Commit(message string) error {
+	return nil
+}
+func (m *mockScm) Clean() error {
+	return nil
+}
+func (m *mockScm) PushTag(tag string) error {
+	return nil
+}
```

Comment:
```suggestion
type mockScm struct {
	scm.Scm

	workingDir string
}

func (m *mockScm) GetDirectory() (directory string) {
	return m.workingDir
}
```

In that case you could `embed` the interface you're trying to stub into your stub structure:
- It saves you defining a bunch of noop methods
- An unexpected call to a non stubbed method results in a panic: because it should not happen and you need to be aware it happened.

When you `embed` something into a structure, if the parent structure doesn't define a method call, then it is implicitely forwarded to the embedded value. 

In our case, let's say I'm calling `Clean` on our mock: 
- it is not defined on the mock, so it  will get delegated to the embedded value
- ...which is an interface, and the zero value of an interface is nil
- ... so it results on a call on a nil value, causing a panic.

embedding is not the most obvious thing to grasp, I would be happy to explain it to you live if you want 😄 

Does it makes sense?

Reply:
That makes 100% sense and is exactly what I was looking for (I mean, this bunch of noop code is annoying to have..). Many thanks!